Errors when trying to setup on MacOS X Sierra

This is the process I followed:
1 - Install MacPorts

2 - `sudo ./prepare-mac-os.sh`

3 - Get a warning: 
`Warning: The Xcode Command Line Tools don't appear to be installed; most ports will likely fail to build.`

Run: `xcode-select --install`

4 - Install dependencies:
`sudo port install autoconf libtool cmake libsdl doxygen readline`

5 - Install automake-1.9 (is this specific version really needed anymore?)
```wget http://ftp.gnu.org/gnu/automake/automake-1.9.tar.bz2
tar jxf automake-1.9.tar.bz2
cd automake-1.9
./configure
make
sudo make install
```

6 - Build and install the toolchain:
`sudo ./toolchain-sudo.sh` 

And then it errors here:
```mkdir psp
mkdir psp/libgcc
Configuring in psp/libgcc
configure: creating cache ./config.cache
checking build system type... x86_64-apple-darwin16.1.0
checking host system type... mipsallegrexel-psp-elf
checking for --enable-version-specific-runtime-libs... no
checking for a BSD-compatible install... /usr/bin/install -c
checking for gawk... awk
checking for psp-ar... /usr/local/pspdev/psp/bin/ar
checking for psp-lipo... psp-lipo
checking for psp-nm... /Users/drm/Code/psptoolchain/build/gcc-4.9.3/build-psp/./gcc/nm
checking for psp-ranlib... /usr/local/pspdev/psp/bin/ranlib
checking for psp-strip... /usr/local/pspdev/psp/bin/strip
checking whether ln -s works... yes
checking for psp-gcc... /Users/drm/Code/psptoolchain/build/gcc-4.9.3/build-psp/./gcc/xgcc -B/Users/drm/Code/psptoolchain/build/gcc-4.9.3/build-psp/./gcc/ -B/usr/local/pspdev/psp/bin/ -B/usr/local/pspdev/psp/lib/ -isystem /usr/local/pspdev/psp/include -isystem /usr/local/pspdev/psp/sys-include   
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether /Users/drm/Code/psptoolchain/build/gcc-4.9.3/build-psp/./gcc/xgcc -B/Users/drm/Code/psptoolchain/build/gcc-4.9.3/build-psp/./gcc/ -B/usr/local/pspdev/psp/bin/ -B/usr/local/pspdev/psp/lib/ -isystem /usr/local/pspdev/psp/include -isystem /usr/local/pspdev/psp/sys-include    accepts -g... yes
checking for /Users/drm/Code/psptoolchain/build/gcc-4.9.3/build-psp/./gcc/xgcc -B/Users/drm/Code/psptoolchain/build/gcc-4.9.3/build-psp/./gcc/ -B/usr/local/pspdev/psp/bin/ -B/usr/local/pspdev/psp/lib/ -isystem /usr/local/pspdev/psp/include -isystem /usr/local/pspdev/psp/sys-include    option to accept ISO C89... unsupported
checking how to run the C preprocessor... /Users/drm/Code/psptoolchain/build/gcc-4.9.3/build-psp/./gcc/xgcc -B/Users/drm/Code/psptoolchain/build/gcc-4.9.3/build-psp/./gcc/ -B/usr/local/pspdev/psp/bin/ -B/usr/local/pspdev/psp/lib/ -isystem /usr/local/pspdev/psp/include -isystem /usr/local/pspdev/psp/sys-include    -E
checking size of double... 8
checking size of long double... 8
checking whether decimal floating point is supported... no
configure: WARNING: decimal float is not supported for this target, ignored
checking whether fixed-point is supported... yes
checking whether to use setjmp/longjmp exceptions... no
checking if the linker (/Users/drm/Code/psptoolchain/build/gcc-4.9.3/build-psp/./gcc/collect-ld) is GNU ld... yes
checking for thread model used by GCC... single
checking whether assembler supports CFI directives... yes
checking whether the target is hard-float... yes
checking for __attribute__((visibility("hidden")))... yes
checking whether the target assembler supports thread-local storage... yes
checking whether the thread-local storage support is from emutls... no
checking for init priority support... yes
configure: updating cache ./config.cache
configure: creating ./config.status
config.status: creating Makefile
config.status: creating auto-target.h
config.status: linking ../../../libgcc/enable-execute-stack-empty.c to enable-execute-stack.c
config.status: linking ../../../libgcc/unwind-generic.h to unwind.h
config.status: linking ../../../libgcc/config/no-unwind.h to md-unwind-support.h
config.status: linking ../../../libgcc/config/mips/sfp-machine.h to sfp-machine.h
config.status: linking ../../../libgcc/gthr-single.h to gthr-default.h
config.status: executing default commands
Adding multilib support to Makefile in ../../../libgcc
multidirs=
with_multisubdir=
# If this is the top-level multilib, build all the other
DEFINES='MIPS_ISA_DEFAULT=2 MIPS_CPU_STRING_DEFAULT=\"allegrex\" MIPS_ABI_DEFAULT=ABI_EABI' HEADERS='../../../libgcc/config/mips/psp.h' \
		../../../libgcc/mkheader.sh > tmp-libgcc_tm.h
# multilibs.
dest=../.././gcc/include/tmp$$-unwind.h; \
	cp unwind.h $dest; \
	chmod a+r $dest; \
	sh ../../../libgcc/../move-if-change $dest ../.././gcc/include/unwind.h
/bin/sh ../../../libgcc/../move-if-change tmp-libgcc_tm.h libgcc_tm.h
echo timestamp > libgcc_tm.stamp
Mode = sf\|df
Suffix = si\|2\|3
Mode = sf\|df
/Users/drm/Code/psptoolchain/build/gcc-4.9.3/build-psp/./gcc/xgcc -B/Users/drm/Code/psptoolchain/build/gcc-4.9.3/build-psp/./gcc/ -B/usr/local/pspdev/psp/bin/ -B/usr/local/pspdev/psp/lib/ -isystem /usr/local/pspdev/psp/include -isystem /usr/local/pspdev/psp/sys-include    -g -O2 -I/opt/local/include -O2  -g -O2  -I/opt/local/include -DIN_GCC  -DCROSS_DIRECTORY_STRUCTURE  -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wstrict-prototypes -Wmissing-prototypes -Wold-style-definition  -isystem ./include   -g -DIN_LIBGCC2 -fbuilding-libgcc -fno-stack-protector -Dinhibit_libc  -I. -I. -I../.././gcc -I../../../libgcc -I../../../libgcc/. -I../../../libgcc/../gcc -I../../../libgcc/../include  -DHAVE_CC_TLS  -o addsf3.o -MT addsf3.o -MD -MP -MF addsf3.dep addsf3 -c ../../../libgcc/config/hardfp.c -fvisibility=hidden -DHIDE_EXPORTS -Wno-missing-prototypes
Suffix = si\|2\|3
/Users/drm/Code/psptoolchain/build/gcc-4.9.3/build-psp/./gcc/xgcc -B/Users/drm/Code/psptoolchain/build/gcc-4.9.3/build-psp/./gcc/ -B/usr/local/pspdev/psp/bin/ -B/usr/local/pspdev/psp/lib/ -isystem /usr/local/pspdev/psp/include -isystem /usr/local/pspdev/psp/sys-include    -g -O2 -I/opt/local/include -O2  -g -O2  -I/opt/local/include -DIN_GCC  -DCROSS_DIRECTORY_STRUCTURE  -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wstrict-prototypes -Wmissing-prototypes -Wold-style-definition  -isystem ./include   -g -DIN_LIBGCC2 -fbuilding-libgcc -fno-stack-protector -Dinhibit_libc  -I. -I. -I../.././gcc -I../../../libgcc -I../../../libgcc/. -I../../../libgcc/../gcc -I../../../libgcc/../include  -DHAVE_CC_TLS  -o subsf3.o -MT subsf3.o -MD -MP -MF subsf3.dep subsf3 -c ../../../libgcc/config/hardfp.c -fvisibility=hidden -DHIDE_EXPORTS -Wno-missing-prototypes
xgcc: error: addsf3: No such file or directory
make[2]: xgcc: error: subsf3: No such file or directory
*** [addsf3.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [subsf3.o] Error 1
if [ -z "" ]; then \
	  true; \
	else \
	  rootpre=`${PWDCMD-pwd}`/; export rootpre; \
	  srcrootpre=`cd ../../../libgcc; ${PWDCMD-pwd}`/; export srcrootpre; \
	  lib=`echo "${rootpre}" | sed -e 's,^.*/\([^/][^/]*\)/$,\1,'`; \
	  compiler="/Users/drm/Code/psptoolchain/build/gcc-4.9.3/build-psp/./gcc/xgcc -B/Users/drm/Code/psptoolchain/build/gcc-4.9.3/build-psp/./gcc/ -B/usr/local/pspdev/psp/bin/ -B/usr/local/pspdev/psp/lib/ -isystem /usr/local/pspdev/psp/include -isystem /usr/local/pspdev/psp/sys-include   "; \
	  for i in `${compiler} --print-multi-lib 2>/dev/null`; do \
	    dir=`echo $i | sed -e 's/;.*$//'`; \
	    if [ "${dir}" = "." ]; then \
	      true; \
	    else \
	      if [ -d ../${dir}/${lib} ]; then \
		flags=`echo $i | sed -e 's/^[^;]*;//' -e 's/@/ -/g'`; \
		if (cd ../${dir}/${lib}; /Applications/Xcode.app/Contents/Developer/usr/bin/make "AR=/usr/local/pspdev/psp/bin/ar" "AR_FLAGS=rc" "CC=/Users/drm/Code/psptoolchain/build/gcc-4.9.3/build-psp/./gcc/xgcc -B/Users/drm/Code/psptoolchain/build/gcc-4.9.3/build-psp/./gcc/ -B/usr/local/pspdev/psp/bin/ -B/usr/local/pspdev/psp/lib/ -isystem /usr/local/pspdev/psp/include -isystem /usr/local/pspdev/psp/sys-include   " "CFLAGS=-g -O2 -I/opt/local/include" "DESTDIR=" "EXTRA_OFILES=" "HDEFINES=" "INSTALL=/usr/bin/install -c" "INSTALL_DATA=/usr/bin/install -c -m 644" "INSTALL_PROGRAM=/usr/bin/install -c" "LDFLAGS=" "LOADLIBES=" "RANLIB=/usr/local/pspdev/psp/bin/ranlib" "SHELL=/bin/sh" "prefix=/usr/local/pspdev" "exec_prefix=/usr/local/pspdev" "libdir=/usr/local/pspdev/lib" "libsubdir=/usr/local/pspdev/lib/gcc/psp/4.9.3" "tooldir=/usr/local/pspdev/psp" \
				CFLAGS="-g -O2 -I/opt/local/include ${flags}" \
				CCASFLAGS=" ${flags}" \
				FCFLAGS=" ${flags}" \
				FFLAGS=" ${flags}" \
				ADAFLAGS=" ${flags}" \
				prefix="/usr/local/pspdev" \
				exec_prefix="/usr/local/pspdev" \
				GCJFLAGS=" ${flags}" \
				GOCFLAGS="-O2 -g ${flags}" \
				CXXFLAGS="-g -O2 ${flags}" \
				LIBCFLAGS="-g -O2  -I/opt/local/include ${flags}" \
				LIBCXXFLAGS="-g -O2 -fno-implicit-templates ${flags}" \
				LDFLAGS=" ${flags}" \
				MULTIFLAGS="${flags}" \
				DESTDIR="" \
				INSTALL="/usr/bin/install -c" \
				INSTALL_DATA="/usr/bin/install -c -m 644" \
				INSTALL_PROGRAM="/usr/bin/install -c" \
				INSTALL_SCRIPT="/usr/bin/install -c" \
				all); then \
		  true; \
		else \
		  exit 1; \
		fi; \
	      else true; \
	      fi; \
	    fi; \
	  done; \
	fi
make[1]: *** [all-target-libgcc] Error 2
make: *** [all] Error 2
../scripts/002-gcc-stage1.sh: Failed.
ERROR: Could not run the toolchain script.
```

Websites that mention the issue:
http://wololo.net/talk/viewtopic.php?f=5&t=45235
http://psx-scene.com/forums/f19/completiing-official-support-ee-gcc5-154680/index2.html

So the `xgcc: error: addsf3: No such file or directory` error turned out to be due to the BSD sed, as highlighted by carstene1ns on #pspdev @ irc.freenode.net.

To get around this:
```
sudo port install gsed
sudo ln -s /opt/local/bin/gsed /usr/local/bin/sed
```

I advise removing the symlink after you're done, since most things on the Mac will assume the BSD sed is the active one.

```
sudo unlink /usr/local/bin/sed
hash -r
```

Side note: I'm not using clang - but instead gcc4.9 (look at my comment above to see how to switch to it using MacPorts).

Anyways, with the GNU sed being used, it got further but failed here:
```
Cloning into 'pspsdk'...
The authenticity of host 'github.com (192.30.253.113)' can't be established.
Are you sure you want to continue connecting (yes/no)? yes
Warning: Permanently added 'github.com,192.30.253.113' (RSA) to the list of known hosts.
Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
...
Making install-data in base
test -z "/usr/local/pspdev/psp/sdk/lib" || /Users/drm/Code/psptoolchain/build/pspsdk/install-sh -d "/usr/local/pspdev/psp/sdk/lib"
test -z "/usr/local/pspdev/psp/sdk/lib" || /Users/drm/Code/psptoolchain/build/pspsdk/install-sh -d "/usr/local/pspdev/psp/sdk/lib"
test -z "/usr/local/pspdev/psp/sdk/include" || /Users/drm/Code/psptoolchain/build/pspsdk/install-sh -d "/usr/local/pspdev/psp/sdk/include"
mkdir: /usr/local/pspdev/psp/sdk/lib: File exists
make[2]: *** [install-buildmakDATA] Error 1
make[2]: *** Waiting for unfinished jobs....
 /usr/bin/install -c -m 644 'as_reg_compat.h' '/usr/local/pspdev/psp/sdk/include/as_reg_compat.h'
 /usr/bin/install -c -m 644 'build_prx.mak' '/usr/local/pspdev/psp/sdk/lib/build_prx.mak'
 /usr/bin/install -c -m 644 'psptypes.h' '/usr/local/pspdev/psp/sdk/include/psptypes.h'
 /usr/bin/install -c -m 644 'linkfile.prx' '/usr/local/pspdev/psp/sdk/lib/linkfile.prx'
 /usr/bin/install -c -m 644 'pspstub.s' '/usr/local/pspdev/psp/sdk/include/pspstub.s'
 /usr/bin/install -c -m 644 'pspimport.s' '/usr/local/pspdev/psp/sdk/include/pspimport.s'
make[1]: *** [install-data-recursive] Error 1
make: *** [install-data-recursive] Error 1
../scripts/003-pspsdk-stage1.sh: Failed.
ERROR: Could not run the toolchain script.
```

Looking into that SSH failure now.

I checked, and yes `/usr/local/pspdev/psp/sdk/lib` was a dir. 
And I was also deleting the build folder after changing stuff + the /usr/local/pspdev folder.

But turns out the issue was that when xgcc was getting built, it was getting built as a Mac App, and Mac Apps are actually folders with meta data. So when things tried to access it as if it were a bin tool - wham.

Things that I have no idea (at the moment) if they had any effect but noting them down:
1 - Added `/opt/local/bin` to the end of `/etc/paths`
2 - I installed libtool (not sure if was already installed - but noting it down just in case)

```
 sudo port install libtool
```

The main thing that I did: 
I installed gnu make and created symlinks for it to override the XCode make.
XCode uses gnu make as well looks like according to the `--version` info, but probably a highly modified version.

```
sudo port install gmake
sudo ln -s /opt/local/bin/gmake /usr/local/bin/make
hash -r
```

To undo the above,
```
sudo unlink /usr/local/bin/make
hash -r
```

It finished with this:
```
Installation finished.
Failed installing: SDL SDL_gfx SDL_image SDL_mixer SDL_ttf bzip2 freetype libmikmod libogg libtremor libvorbis opentri smpeg-psp
Remember to add /usr/local/pspdev/bin to your path...
```

It got most of it done, except the aforementioned failed installations.

I'll delete everything and start from scratch to try to find out what's necessary and what's not.
But before I do that, here are some of the errors from this build (that I plan to look into probably this weekend, but sharing them for those interested):
```
Connecting to www.bzip.org (www.bzip.org)|46.235.226.80|:80... connected.
HTTP request sent, awaiting response... 206 Partial Content
Length: 782025 (764K), 454867 (444K) remaining [application/x-gzip]
Saving to: ‘bzip2-1.0.6.tar.gz’

bzip2-1.0.6.tar.gz                              100%[++++++++++++++++++++++++++++++++++++++++++===========================================================>] 763.70K   301KB/s    in 1.5s    

2017-01-25 01:20:04 (301 KB/s) - ‘bzip2-1.0.6.tar.gz’ saved [782025/782025]

Extracting bzip2-1.0.6.tar.gz...
tar: Unrecognized archive format
tar: Error exit delayed from previous errors.
Failed installing bzip2!
```

```
./src/video/psp/SDL_pspevents.c:42:23: fatal error: pspirkeyb.h: No such file or directory
 #include <pspirkeyb.h>
                       ^
compilation terminated.
build-deps:910: recipe for target 'build/SDL_pspevents.lo' failed
make: *** [build/SDL_pspevents.lo] Error 1
make: *** Waiting for unfinished jobs....
Failed installing SDL!
```

```
MPEGfilter.c:22:17: fatal error: SDL.h: No such file or directory
 #include "SDL.h"
                 ^
compilation terminated.
```

I assume the below one is just to install libtoolize...
```
running `libtoolize --copy'
autogen.sh: line 15: libtoolize: command not found
error while running `libtoolize --copy'
Couldn't install dependency freetype, aborting.
Failed installing opentri!
```

```
configure.in:190: error: possibly undefined macro: AM_PROG_LIBTOOL
      If this token and others are legitimate, please use m4_pattern_allow.
      See the Autoconf documentation.
configure.in:192: error: possibly undefined macro: AC_PROG_LIBTOOL
Failed installing libmikmod!
```

```
configure.in:201: the top level
src/Makefile.am:5: Libtool library used but `LIBTOOL' is undefined
src/Makefile.am:5: 
src/Makefile.am:5: The usual way to define `LIBTOOL' is to add `AC_PROG_LIBTOOL'
src/Makefile.am:5: to `configure.in' and run `aclocal' and `autoconf' again.
Failed installing libogg!
```

```
Makefile.am:5: Libtool library used but `LIBTOOL' is undefined
Makefile.am:5: 
Makefile.am:5: The usual way to define `LIBTOOL' is to add `AC_PROG_LIBTOOL'
Makefile.am:5: to `configure.in' and run `aclocal' and `autoconf' again.
Makefile.am: installing `./depcomp'
Failed installing libtremor!
```

```
Useless use of /d modifier in transliteration operator at /usr/local/share/automake-1.9/Automake/Wrap.pm line 60.
lib/Makefile.am:7: Libtool library used but `LIBTOOL' is undefined
lib/Makefile.am:7: 
lib/Makefile.am:7: The usual way to define `LIBTOOL' is to add `AC_PROG_LIBTOOL'
lib/Makefile.am:7: to `configure.in' and run `aclocal' and `autoconf' again.
Failed installing libvorbis!
```

```
MPEGring.cpp:26:23: fatal error: SDL_timer.h: No such file or directory
 #include "SDL_timer.h"
                       ^
compilation terminated.
<builtin>: recipe for target 'MPEGring.o' failed
make: *** [MPEGring.o] Error 1
Failed installing smpeg-psp!
```

I wouldn't be surprised if most of these errors stem from SDL failing.

Oh yeah, about enforcing gnu gcc:

Looks like in `./prepare-mac-os.sh`, libusb-legacy needs to be built with clang.
I was getting errors about `-no-cpp-precomp` using gnu gcc.

So the solution should require clang for the setup part and gcc for the toolchain part.

I'll probably take on doing the PR for this in a weekend or two.

Finally had the free time to look into this more.

Most of the errors went away once I added in libtoolize + let port install the latest automake (1.15 at-this-time) instead of using automake-1.9.

The preliminary patch:

```
diff --git a/prepare-mac-os.sh b/prepare-mac-os.sh
index 2f7fe83..bbe874b 100755
--- a/prepare-mac-os.sh
+++ b/prepare-mac-os.sh
@@ -10,12 +10,44 @@ install_libusb() {
 }
 
 if [ -e "/opt/local/bin/port" ]; then
-       sudo port install wget
-       sudo port install libelf
-       sudo port install libmpc
-       sudo port install libusb
-       
+       if [[ $UID != 0 ]]; then
+           echo "Please run this script with sudo"
+           exit 1
+       fi
+
+       port install wget
+       port install libelf
+       port install libmpc
+       port install libusb
+       port install autoconf
+       port install automake
+       port install libtool
+       port install cmake
+       port install libsdl
+       port install doxygen
+       port install readline
+
+       port install gcc49
+       port install gcc_select
+
+       # Build with clang
+       port select --set gcc none
+
        install_libusb
+
+       # Switch to GNU variants
+       mkdir gnubin
+
+       port select --set gcc mp-gcc49
+
+       port install gsed
+       ln -s /opt/local/bin/gsed gnubin/sed
+
+       port install gmake
+       ln -s /opt/local/bin/gmake gnubin/make
+
+       port install libtool
+       ln -s /opt/local/bin/glibtoolize gnubin/libtoolize
+
 elif [ -e "/usr/local/bin/brew" ]; then
        CURRENT_USER=$(stat -f '%Su' /dev/console)
        sudo -u $CURRENT_USER brew install wget
diff --git a/toolchain-sudo.sh b/toolchain-sudo.sh
index 389faa4..71f3eed 100755
--- a/toolchain-sudo.sh
+++ b/toolchain-sudo.sh
@@ -1,5 +1,10 @@
 #!/bin/bash
 # toolchain-sudo.sh by Dan Peori (danpeori@oopo.net)
+
+ if [[ $UID != 0 ]]; then
+   echo "Please run this script with sudo"
+   exit 1
+ fi
  
  INSTALLDIR=/usr/local/pspdev
  
@@ -8,7 +13,7 @@
 
  ## Set up the environment.
  export PSPDEV=$INSTALLDIR
- export PATH=$PATH:$PSPDEV/bin
+ export PATH=`pwd`/gnubin:$PATH:$PSPDEV/bin
 
  ## Run the toolchain script.
  ./toolchain.sh $@ || { echo "ERROR: Could not run the toolchain script."; exit 1; }
```

Two problems remain:

1 - On the first run, the following still fail:
```
Installation finished.
Failed installing: SDL SDL_gfx SDL_image SDL_mixer SDL_ttf opentri smpeg-psp
```

The error for all of them (except opentri) is:
```
./src/video/psp/SDL_pspevents.c:42:23: fatal error: pspirkeyb.h: No such file or directory
 #include <pspirkeyb.h>
```

So turns out the missing file pspirkeyb.h is placed much later after SDL. 
For me, said file is copied at line 64222 in my stdout/stderr log:
```
cp include/pspirkeyb.h /usr/local/pspdev/psp/sdk/include
```
While the SDL error I posted above is at line 55487 (much earlier) in the same said log.

If I don't clean /usr/local/pspdev/ when I re-run `sudo ./toolchain-sudo.sh`, SDL and the others get built. 

Of course, telling people to re-run the script twice is no solution ...

Still not sure why this is happening, still looking into it.
I wonder if it's some multi-threading gone awry.

2 - While number 1 is due to an out-of-order issue, the only other remaining failure is:
```
Installation finished.
Failed installing: opentri
``` 

This is the same issue as mentioned in issue #72 
```
psp-gcc -I. -I/usr/local/pspdev/psp/sdk/include -O3 -G0 -Wall -D__PSP__ -DTRI_SUPPORT_PNG -DTRI_SUPPORT_FT -I/usr/local/pspdev/psp/include/freetype2 -I/usr/local/pspdev/psp/include -D_PSP_FW_VERSION=150   -c -o triCamera.o triCamera.c
triArchive.c: In function ‘triArchiveRead’:
triArchive.c:131:11: warning: suggest parentheses around comparison in operand of ‘&’ [-Wparentheses]
  if(flags & TRI_ARCHIVE_ENC == TRI_ARCHIVE_ENC)^M
           ^
triArchive.c:142:11: warning: suggest parentheses around comparison in operand of ‘&’ [-Wparentheses]
  if(flags & TRI_ARCHIVE_COMP == TRI_ARCHIVE_COMP)^M
           ^
triArchive.c:145:15: warning: suggest parentheses around comparison in operand of ‘&’ [-Wparentheses]
   if(comptype & TRI_ARCHIVE_ZLIB == TRI_ARCHIVE_ZLIB)^M
               ^
streams/streams.c: In function ‘memcpy_vfpu’:
streams/streams.c:93:15: warning: initialization makes pointer from integer without a cast
  u8* dst64a = ((u32)dst8&~0x3F);^M
               ^
triTexman.c:537:15: warning: ‘triTextureManageUploads’ defined but not used [-Wunused-function]
 static triS32 triTextureManageUploads()^M
               ^
triTexman.c: In function ‘triNearestColorIndex’:
triTexman.c:820:1: error: unrecognizable insn:
 }^M
 ^
(insn 237 236 238 37 (set (reg:SI 477)
        (if_then_else:SI (ne:CC (reg:CC 67 $fcc0)
                (const_int 0 [0]))
            (reg/v:SI 281 [ minIndex ])
            (reg/v:SI 265 [ minIndex ]))) -1
     (nil))
triTexman.c:820:1: internal compiler error: in extract_insn, at recog.c:2202

triTexman.c:820:1: internal compiler error: Abort trap: 6
psp-gcc: internal compiler error: Abort trap: 6 (program cc1)
make: *** [<builtin>: triTexman.o] Abort trap: 6
```

Not sure about this one either.

The opentri bug looks to be something to do with this: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=58158

$ cat reproducer.c
```
extern void foo(void);

static inline float max(const float x, const float y)
{
        return (x > y) ? x : y;
}

void bar(int d)
{
        int i;

        for (i = 0; i < (int)max(100.f * d, 1.f); i++) {
                foo();
        }
}
```

$ /usr/local/pspdev/bin/psp-gcc -O2 -c -o reproducer.o reproducer.c
```
reproducer.c: In function ‘bar’:
reproducer.c:15:1: error: unrecognizable insn:
 }
 ^
(insn 19 18 20 2 (set (reg:SI 202 [ D.1448 ])
        (if_then_else:SI (ne:CC (reg:CC 67 $fcc0)
                (const_int 0 [0]))
            (reg:SI 202 [ D.1448 ])
            (reg:SI 204))) -1
     (nil))
reproducer.c:15:1: internal compiler error: in extract_insn, at recog.c:2202

reproducer.c:15:1: internal compiler error: Abort trap: 6
psp-gcc: internal compiler error: Abort trap: 6 (program cc1)
Abort trap: 6
```

It's listed as known to fail for 4.9.4, while comments states it should have been fixed for 4.9.3.
It's stated to work for 5.0.

--- a/README.md
+++ b/README.md
@@ -1,61 +1,77 @@
 What does this do?
 ==================
 
-This program will automatically build and install a compiler and other tools used in the creation of homebrew software for the Sony Playstation Portable handheld videogame system.
+This program will automatically build and install a compiler and other tools
+used in the creation of homebrew software for the Sony Playstation Portable
+handheld videogame system.
 
 How do I use it?
 ==================
 
 1. Set up your environment by installing the following software:
 
-        autoconf, automake, bison, bzip2, flex, gcc, g++/gcc-c++, gzip, libusb-dev, make, ncurses, patch, readline, subversion, texinfo, xz-utils, wget, libelf, git
+        autoconf, automake, bison, bzip2, flex, gcc, g++/gcc-c++, gzip,
+        libusb-dev, make, ncurses, patch, readline, subversion, texinfo,
+        xz-utils, wget, libelf, git
 
 2. Set the PSPDEV and PATH environmental variables:
 
-        export PSPDEV=/usr/local/pspdev
-        export PATH=$PATH:$PSPDEV/bin
+    ```shell
+    export PSPDEV=/usr/local/pspdev
+    export PATH=$PATH:$PSPDEV/bin
+    ```
 
-    The PSPDEV variable is the directory the toolchain will be installed to, change this if you wish. If possible the toolchain script will automatically add these variables to your systems login scripts, otherwise you will need to manually add these variables yourself.
+    The PSPDEV variable is the directory the toolchain will be installed to,
+    change this if you wish. If possible the toolchain script will automatically
+    add these variables to your systems login scripts, otherwise you will need
+    to manually add these variables yourself.
 
 3. Run the toolchain script:
 
         ./toolchain.sh
 
-> NOTE: If you have issues with compiling try increasing the amount of memory
-> available to your system by creating a swapfile.
->
->     dd if=/dev/zero of=/swapfile bs=1M count=2048
->     chmod 600 /swapfile
->     mkswap /swapfile
->     swapon /swapfile
->
-> After you are done use `swapoff -a` disable the swapfile. Finally you can
-> remove it with `rm`.
+> NOTE: If you have issues with compiling try increasing the amount of
+        memory available to your system by creating a swapfile.
+
+```shell
+dd if=/dev/zero of=/swapfile bs=1M count=2048
+chmod 600 /swapfile
+mkswap /swapfile
+swapon /swapfile
+[...compilation...]
+swapoff /swapfile
+rm /swapfile
+```
 
 Ubuntu/Debian
 -------------
 
 1. Install the required packages by running:
 
-    sudo ./prepare-debian-ubuntu.sh
+        sudo ./prepare-debian-ubuntu.sh
 
 2. Build and install the toolchain and SDK.
 
-    sudo ./toolchain-sudo.sh
+        sudo ./toolchain-sudo.sh
  
-> NOTE: If you do not wish for the toolchain to be installed in /usr/local/pspdev then edit toolchain-sudo.sh and change the INSTALLDIR variable.
+    > NOTE: If you do not wish for the toolchain to be installed in
+            `/usr/local/pspdev` then edit toolchain-sudo.sh and change the
+            `$INSTALLDIR` variable.
 
 OSX
 ---
 
-1. Install [`port`][MacPorts] or [`brew`][HomeBrew].
+1. Install [MacPorts][MacPorts] or [Homebrew][Homebrew].
 2. Install all the libraries you will need before building by running:
         
-    sudo ./prepare-mac-os.sh
+        sudo ./prepare-mac-os.sh
+
+    > NOTE: You can use `-b` or `--brew` flag for using Homebrew and
+            `-p` or `--port` flag for using MacPorts.
 
 3. Build and install the toolchain and SDK.
         
-    sudo ./toolchain-sudo.sh
+        sudo ./toolchain-sudo.sh
 
 Where do I go from here?
 ========================

--- a/depends/check-dependencies.sh
+++ b/depends/check-dependencies.sh
@@ -12,11 +12,6 @@ header_paths=(
     # -- Add more locations here --
 )
 
-aclocal_paths=(
-    "/usr/share/aclocal"
-    "/usr/local/share/aclocal"
-)
-
 missing_depends=()
 
 function check_header
@@ -53,14 +48,10 @@ function check_program
     missing_depends+=($1); return 1
 }
 
-function check_aclocal
-{
-    for place in ${aclocal_paths[@]}; do
-        [ -f "$place/$2" ] && return 0
-    done
-    
-    missing_depends+=($1); return 1
-}
+# macOS catalina does not ship headers in default directory anymore
+if [ "$(uname)" == "Darwin" ]; then
+  header_paths+=("`xcrun --show-sdk-path`/usr/include")
+fi
 
 check_header    libelf          elf.h libelf.h libelf/libelf.h gelf.h libelf/gelf.h
 check_header    libusb          usb.h

--- a/prepare-mac-os.sh
+++ b/prepare-mac-os.sh
@@ -1,5 +1,6 @@
 #!/bin/sh
 
+# only needed for macports
 install_libusb() {
 	wget --continue https://downloads.sourceforge.net/project/libusb/libusb-0.1%20%28LEGACY%29/0.1.12/libusb-0.1.12.tar.gz -O libusb-legacy.tar.gz
 	rm -Rf libusb-legacy && mkdir libusb-legacy && tar --strip-components=1 --directory=libusb-legacy -xzf libusb-legacy.tar.gz
@@ -9,13 +10,59 @@ install_libusb() {
 	rm -Rf libusb-legacy && rm libusb-legacy.tar.gz
 }
 
-if [ -e "/opt/local/bin/port" ]; then
-	sudo port install autoconf automake cmake doxygen gsed libelf libtool libusb pkgconfig wget xz
-	
-	install_libusb
-elif [ -e "/usr/local/bin/brew" ]; then
+show_help() {
+	echo "Use '-b' or '--brew' to install packages with Homebrew, or use '-p' or '--port' to install with MacPorts."
+	echo "If left unspecified, tries Homebrew first, then MacPorts."
+	exit 1
+}
+
+if [ $# -gt 1 ]; then
+	echo "Invalid number of command line options."
+	show_help
+fi
+
+# default is autodetection
+try_brew=1
+try_port=1
+if [ $# -eq 1 ]; then
+	case "$1" in
+	-b|--brew)
+		try_port=0 ;;
+	-p|--port)
+		try_brew=0 ;;
+	*)
+		show_help ;;
+	esac
+fi
+
+if ! [ -e "/usr/local/bin/brew" -o -e "/opt/local/bin/port" ]; then
+	echo "Go install Homebrew from http://brew.sh/ or MacPorts from http://www.macports.org/ first, then we can talk!"
+	exit 1
+fi
+
+# sanity checks
+if [ $try_brew -eq 1 -a ! -e "/usr/local/bin/brew" ]; then
+	echo "Not trying Homebrew, because it is not installed."
+	try_brew=0
+fi
+if [ $try_port -eq 1 -a ! -e "/opt/local/bin/port" ]; then
+	echo "Not trying MacPorts, because it is not installed."
+	try_port=0
+fi
+if [ $try_brew -eq 0 -a $try_port -eq 0 ]; then
+	echo "Invalid package manager specified. Maybe use autodetection."
+	exit 1
+fi
+
+# actual installation
+if [ $try_brew -eq 1 ]; then
 	CURRENT_USER=$(stat -f '%Su' /dev/console)
 	sudo -u $CURRENT_USER brew install autoconf automake cmake doxygen gnu-sed libelf libool libusb libusb-compat pkg-config wget xz
-else
-	echo "Go install MacPorts from http://www.macports.org/ or Homebrew from http://brew.sh/ first, then we can talk"
+	exit
+fi
+if [ $try_port -eq 1 ]; then
+	sudo port install autoconf automake cmake doxygen gsed libelf libtool libusb pkgconfig wget xz
+
+	install_libusb
+	exit
 fi

--- a/prepare-mac-os.sh
+++ b/prepare-mac-os.sh
@@ -10,12 +10,12 @@ install_libusb() {
 }
 
 if [ -e "/opt/local/bin/port" ]; then
-	sudo port install automake cmake doxygen gsed libelf libusb wget xz
+	sudo port install autoconf automake cmake doxygen gsed libelf libtool libusb pkgconfig wget xz
 	
 	install_libusb
 elif [ -e "/usr/local/bin/brew" ]; then
 	CURRENT_USER=$(stat -f '%Su' /dev/console)
-	sudo -u $CURRENT_USER brew install automake cmake doxygen gnu-sed libelf libusb libusb-compat wget xz
+	sudo -u $CURRENT_USER brew install autoconf automake cmake doxygen gnu-sed libelf libool libusb libusb-compat pkg-config wget xz
 else
 	echo "Go install MacPorts from http://www.macports.org/ or Homebrew from http://brew.sh/ first, then we can talk"
 fi

--- a/prepare-mac-os.sh
+++ b/prepare-mac-os.sh
@@ -10,12 +10,12 @@ install_libusb() {
 }
 
 if [ -e "/opt/local/bin/port" ]; then
-	sudo port install automake cmake doxygen libelf libusb wget xz
+	sudo port install automake cmake doxygen gsed libelf libusb wget xz
 	
 	install_libusb
 elif [ -e "/usr/local/bin/brew" ]; then
 	CURRENT_USER=$(stat -f '%Su' /dev/console)
-	sudo -u $CURRENT_USER brew install automake cmake doxygen libelf libusb libusb-compat wget xz
+	sudo -u $CURRENT_USER brew install automake cmake doxygen gnu-sed libelf libusb libusb-compat wget xz
 else
 	echo "Go install MacPorts from http://www.macports.org/ or Homebrew from http://brew.sh/ first, then we can talk"
 fi

--- a/scripts/002-gcc-stage1.sh
+++ b/scripts/002-gcc-stage1.sh
@@ -34,6 +34,11 @@
  mkdir build-psp
  cd build-psp
 
+ ## Under macOS we need the gnu variant of sed
+ if [ "$(uname)" == "Darwin" ]; then
+   export PATH="/usr/local/opt/gnu-sed/libexec/gnubin:/opt/local/libexec/gnubin:$PATH"
+ fi
+
  ## Configure the build.
  CFLAGS="$CFLAGS -I/opt/local/include" \
    CPPFLAGS="$CPPFLAGS -I/opt/local/include" \

--- a/scripts/005-gcc-stage2.sh
+++ b/scripts/005-gcc-stage2.sh
@@ -35,6 +35,11 @@
  mkdir build-psp
  cd build-psp
 
+ ## Under macOS we need the gnu variant of sed
+ if [ "$(uname)" == "Darwin" ]; then
+   export PATH="/usr/local/opt/gnu-sed/libexec/gnubin:/opt/local/libexec/gnubin:$PATH"
+ fi
+
  ## Configure the build.
  ## If you want to build gdc add "d" to --enable-languages option.
  CFLAGS="$CFLAGS -I/opt/local/include" \

